src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/RuleBasedTests.cs

```diff
@@ -57,30 +63,103 @@ public static IEnumerable<object[]> GetHL7V2Cases()
+                new object[] { "CCD", "sample.ccda" },
+                new object[] { "CCD", "Transfer_Summary.ccda" },
+                new object[] { "CCD", "Unstructured_Document_embed.ccda" },
+                new object[] { "CCD", "Unstructured_Document_reference.ccda" },
+                new object[] { @"ConsultationNote", @"Care_Plan.ccda"},
+                new object[] { @"ConsultationNote", @"CDA_with_Embedded_PDF.ccda"},
+                new object[] { @"ConsultationNote", @"Consultation_Note.ccda"},
+                new object[] { @"ConsultationNote", @"Unstructured_Document_embed.ccda"},
+                new object[] { @"DischargeSummary", @"Discharge_Summary.ccda"},
+                new object[] { @"DischargeSummary", @"Consult-Document-Closing-Referral-C-CDAR2.1.ccda"},
+                new object[] { @"HistoryandPhysical", @"History_and_Physical.ccda"},
+                new object[] { @"HistoryandPhysical", @"Diagnostic_Imaging_Report.ccda"},
+                new object[] { @"OperativeNote", @"Operative_Note.ccda"},
+                new object[] { @"OperativeNote", @"Patient-1.ccda"},
+                new object[] { @"ProcedureNote", @"Procedure_Note.ccda"},
+                new object[] { @"ProcedureNote", @"Patient-and-Provider-Organization-Direct-Address-C-CDAR2.1.ccda"},
+                new object[] { @"ProgressNote", @"Progress_Note.ccda"},
+                new object[] { @"ProgressNote", @"PROBLEMS_in_Empty_C-CDA_2.1-C-CDAR2.1.ccda"},
+                new object[] { @"ReferralNote", @"Referral_Note.ccda"},
+                new object[] { @"ReferralNote", @"sample.ccda"},
+                new object[] { @"TransferSummary", @"Transfer_Summary.ccda"},
+                new object[] { @"TransferSummary", @"Unstructured_Document_reference.ccda"},
+            };
+            return cases.Select(item => new object[]
+            {
+                Convert.ToString(item[0]),
+                Path.Combine(_ccdaDataFolder, Convert.ToString(item[1])),
+                DataType.Ccda,
+            });
         }
 
         [Theory]
         [MemberData(nameof(GetHL7V2Cases))]
         [MemberData(nameof(GetCcdaCases))]
-        public async Task CheckOnePatient(string templateName, string samplePath)
+        public async Task CheckOnePatient(string templateName, string samplePath, DataType dataType)
         {
-            var result = await ConvertData(templateName, samplePath);
+            var result = await ConvertData(templateName, samplePath, dataType);
             var patients = result.SelectTokens("$.entry[?(@.resource.resourceType == 'Patient')].resource.id");
             Assert.Equal(1, patients?.Count());
         }
 
+        [Theory]
+        [MemberData(nameof(GetCcdaCases))]
+        public async Task CheckReferenceResourceId(string templateName, string samplePath, DataType dataType)
+        {
+            Dictionary<string, string> resourceIdMapping = new Dictionary<string, string>();
```

Comment:
It looks we can use a List<string> to store resource IDs like `{resourceType}/{resourceId}`. Resource IDs only need to be unique within a resource type. For example, we can have `Patient/example` and `Observation/example` at the same time.

Reply:
HashSet is used since the time complexity of `List.Contains` is O(N).